devito/finite_differences/differentiable.py

```diff
@@ -0,0 +1,236 @@
+import sympy
+import numpy as np
+
+from devito.symbolics.search import retrieve_functions
+from devito.symbolics.extended_sympy import FrozenExpr
```

Comment:
I still don't understand why we need this to be a FrozenExpr. I think it shouldn't be. Subclassing ``sympy.Expr`` should be enough.

Reply:
Why do we need the custom xreplace? 

so, here's my take on this. Differentiable is only useful in user-land. Once we're compiling, we could even safely get rid of it. Correct me if I'm wrong.

Reply:
Yes, it's only for symbolic FD/manipulation, not required for the compiler